### PR TITLE
 Add SMN, DAM, DAM-USE-T ranking models

### PR DESCRIPTION
**Pull request by golovenkov**
### Description
Add several ranking matching models for retrieval-based multi-turn response selection:

- Sequential Matching Network (SMN) [Wu et al., 2017]

- Deep Attention Matching Network (DAM) [Zhou et al., 2018]

- Deep Attention Matching Network with Universal Sentence Encoder (DAM-USE-T). This is our new proposed neural ranking architecture based on [Zhou et al., 2018] and [Cer D. et al., 2018].
The model DAM-USE-T **outperforms** the current state-of-the-art models for the Ubuntu Dialogue Corpus v1 [Lowe et al., 2015] with the test metrics: `0.7957/0.8932/0.9734`
The DAM-USE-T has high test metrics for the Ubuntu Dialogue Corpus v2 [Lowe et al., 2017]  as well: `0.7446/0.8677/0.9738`.

Scores for **Ubuntu Dialogue Corpus v1** (test):

| Model | R@1 | R@2 | R@5 |
| --- | --- | --- | --- |
| SMN last [Wu et al., 2017] | 0.723 | 0.842 | 0.956 |
| SMN last (DeepPavlov) | 0.759 | 0.871 | 0.968 |
| DAM [Zhou et al., 2018] | 0.767 | 0.874 | 0.969 |
| DAM (DeepPavlov) | 0.779 | 0.880 | 0.970 |  
| MRFN-FLS [Tao et al., 2019] | 0.786 | 0.886 | 0.976 |
| IMN [Gu et al, 2019] | 0.777 | 0.880 | 0.974 |
| IMN Ensemble [Gu et al, 2019] | 0.794 | 0.893 | **0.978** |
| DAM-USE-T (DeepPavlov) | **0.7957** | **0.8932** | 0.9734 |

Scores for **Ubuntu Dialogue Corpus v2** (test):

| Model | R@1 | R@2 | R@5 |
| --- | --- | --- | --- |
| SMN last [Wu et al., 2017] | - | - | - |
| SMN last (DeepPavlov) | 0.6791 | 0.8149 | 0.9563 |
| DAM [Zhou et al., 2018] | - | - | - |
| DAM (DeepPavlov) | 0.7154 | 0.8366 | 0.9633 |
| MRFN-FLS [Tao et al., 2019] | - | - | - |
| IMN [Gu et al, 2019] | 0.771 | 0.886 | 0.979 |
| IMN Ensemble [Gu et al, 2019] | **0.791** | **0.899** | **0.982** |
| DAM-USE-T (DeepPavlov) | 0.7446 | 0.8677 | 0.9738 |

**UPD**: added pre-trained models (SMN, DAM, DAM-USE-T) for both Ubuntu Corpus v1 and v2 datasets. There are 6 models in total, see ranking configs, for example, pre-trained model DAM-USE-T for Ubuntu-v1 is  [here](https://github.com/deepmipt/DeepPavlov/blob/57f4bbd869047892eae46e1b29eaf5690d5887e7/deeppavlov/configs/ranking/ranking_ubuntu_v1_mt_word2vec_dam_transformer.json#L123-L126)

### References
Ryan Lowe, Nissan Pow, Iulian Serban, and Joelle Pineau. 2015. The ubuntu dialogue corpus: A large dataset for research in unstructured multi-turn dialogue systems. arXiv preprint arXiv:1506.08909.

Ryan Thomas Lowe, Nissan Pow, Iulian Vlad Serban, Laurent Charlin, Chia-Wei Liu, and Joelle Pineau. 2017. Training end-to-end dialogue systems with the ubuntu dialogue corpus. Dialogue and Discourse 8(1).

Xiangyang Zhou, Lu Li, Daxiang Dong, Yi Liu, Ying Chen, Wayne Xin Zhao, Dianhai Yu and Hua Wu. 2018. Multi-Turn Response Selection for Chatbots with Deep Attention Matching Network. Proceedings of the 56th Annual Meeting of the Association for Computational Linguistics (Volume 1: Long Papers), pages 1118-1127, ACL. http://aclweb.org/anthology/P18-1103 

Yu Wu, Wei Wu, Ming Zhou, and Zhoujun Li. 2017. Sequential match network: A new architecture for multi-turn response selection in retrieval-based chatbots. In ACL, pages 372–381.

Daniel Cer, Yinfei Yang, Sheng-yi Kong, Nan Hua, Nicole Limtiaco, Rhomni St. John, Noah Constant, Mario Guajardo-Cespedes, Steve Yuan, Chris Tar, Brian Strope, Ray Kurzweil. 2018a. Universal Sentence Encoder for English. Proceedings of the 2018 Conference on Empirical Methods in Natural Language Processing: System Demonstrations. Pages 169–174. http://aclweb.org/anthology/D18-2029 

Chongyang Tao, Wei Wu, Can Xu, Wenpeng Hu, Dongyan Zhao, and Rui Yan. Multi-Representation Fusion Network for Multi-turn Response Selection in Retrieval-based Chatbots.  In WSDM'19. https://dl.acm.org/citation.cfm?id=3290985

Gu, Jia-Chen & Ling, Zhen-Hua & Liu, Quan. (2019). Interactive Matching Network for Multi-Turn Response Selection in Retrieval-Based Chatbots. https://arxiv.org/abs/1901.01824

### CODE REVIEW

File: deeppavlov/dataset_readers/ubuntu_v1_mt_reader.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+from pathlib import Path
+import codecs
+
+from deeppavlov.core.common.registry import register
+from deeppavlov.core.data.dataset_reader import DatasetReader
+
+
+@register('ubuntu_v1_mt_reader')
+class UbuntuV1MTReader(DatasetReader):
+    """The class to read the Ubuntu V2 dataset from csv files taking into account multi-turn dialogue ``context``.
```
Comment (puleon) [inline deeppavlov/dataset_readers/ubuntu_v1_mt_reader.py]: V1

File: docs/components/neural_ranking.rst [hunk 2]
```diff
@@ -35,6 +35,38 @@ To train from command line:
...
+
+    python -m deeppavlov train -d ranking_ubuntu_v2_mt_word2vec_smn
+    python -m deeppavlov train -d ranking_ubuntu_v2_mt_word2vec_dam
+    python -m deeppavlov train -d ranking_ubuntu_v2_mt_word2vec_dam_transformer
+
+As an example of configuration file see
+:config:`ranking_ubuntu_v2_mt_word2vec_smn.json <ranking/ranking_ubuntu_v2_mt_word2vec_smn.json>`.
+
+
+Paraphrase identification
```
Comment (puleon) [inline docs/components/neural_ranking.rst]:
```suggestion
```

File: docs/intro/features.rst [hunk 3]
```diff
@@ -217,26 +217,56 @@ on Automatic Spelling Correction for Russian:
 :doc:`Ranking component </components/neural_ranking>`
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-Based on `LSTM-based deep learning models for non-factoid answer selection <https://arxiv.org/abs/1511.04108>`__. The
-model performs ranking of responses or contexts from some database by their relevance for the given context.
+The main neural ranking model based on `LSTM-based deep learning models for non-factoid answer selection
+<https://arxiv.org/abs/1511.04108>`__. The model performs ranking of responses or contexts from some database by their
+relevance for the given context.
+
+There are 3 alternative neural architectures available as well:
```
Comment (puleon) [inline docs/intro/features.rst]: Please, add 2 tables from this pull request description to the "comparison with other models" section for the ranking component in this file, add references to models configs from these 2 tables.

File: deeppavlov/dataset_readers/ubuntu_v1_mt_reader.py (hunk 1 above)
Comment (golovenkov) [reply deeppavlov/dataset_readers/ubuntu_v1_mt_reader.py]: fixed

File: docs/components/neural_ranking.rst (hunk 2 above)
Comment (golovenkov) [reply docs/components/neural_ranking.rst]: removed

File: docs/components/neural_ranking.rst [hunk 4]
```diff
@@ -35,6 +35,38 @@ To train from command line:
...
+-------------------------
```
[the remark this answers is not in the document]
Comment (golovenkov) [reply docs/components/neural_ranking.rst]: removed

File: docs/intro/features.rst (hunk 3 above)
Comment (golovenkov) [reply docs/intro/features.rst]: done

File: deeppavlov/dataset_readers/ubuntu_dstc7_mt_reader.py [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import List, Tuple, Dict
+from pathlib import Path
+import json
+import numpy as np
```
Comment (yoptar) [inline deeppavlov/dataset_readers/ubuntu_dstc7_mt_reader.py]:
```suggestion

import numpy as np
```

File: deeppavlov/dataset_readers/ubuntu_dstc7_mt_reader.py [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+            res = sent_list + (self.num_turns - len(sent_list)) * \
+                  [''] if len(sent_list) < self.num_turns else sent_list[:self.num_turns]
+            return res
+        elif padding == "pre":
+            sent_list = context[-(self.num_turns + 1):-1]
+            if len(sent_list) <= self.num_turns:
+                tmp = sent_list[:]
+                sent_list = [''] * (self.num_turns - len(sent_list))
+                sent_list.extend(tmp)
+            return sent_list
```
Comment (yoptar) [inline deeppavlov/dataset_readers/ubuntu_dstc7_mt_reader.py]:
```suggestion
            return sent_list

```

File: deeppavlov/dataset_readers/ubuntu_v1_mt_reader.py [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+        dataset["train"] = self.preprocess_data_train(train_fname)
+        dataset["valid"] = self.preprocess_data_validation(valid_fname)
+        dataset["test"] = self.preprocess_data_validation(test_fname)
+        return dataset
+
+    def preprocess_data_train(self, train_fname: Union[Path, str]) -> List[Tuple[List[str], int]]:
+        contexts = []
+        responses = []
+        labels = []
+        with codecs.open(train_fname, 'r', 'utf-8') as f:
```
Comment (yoptar) [inline deeppavlov/dataset_readers/ubuntu_v1_mt_reader.py]: why not just `open(train_fname, encoding='utf-8')`?

File: deeppavlov/models/preprocessors/siamese_preprocessor.py [hunk 8]
```diff
@@ -70,10 +72,11 @@ def __init__(self,
                  use_matrix: bool = True,
                  num_context_turns: int = 1,
                  num_ranking_samples: int = 1,
+                 add_raw_text: bool = False,
                  tokenizer: Component = None,
-                 vocab: Estimator = "simple_vocab",
-                 embedder: Component = "fasttext",
-                 sent_vocab: Estimator = "simple_vocab",
+                 vocab: Estimator = None,
```
Comment (yoptar) [inline deeppavlov/models/preprocessors/siamese_preprocessor.py]:
```suggestion
                 vocab: Optional[Estimator] = None,
```

File: deeppavlov/models/ranking/deep_attention_matching_network.py [hunk 9]
```diff
@@ -0,0 +1,246 @@
...
+                 filters2_conv3d: int = 16,
+                 stack_num: int = 5,
+                 seed: int = 65,
+                 decay_steps: int = 600,
+                 *args,
+                 **kwargs):
+
+        self.seed = seed
+
+        np.random.seed(self.seed)
```
Comment (yoptar) [inline deeppavlov/models/ranking/deep_attention_matching_network.py]: Looks like `np.random` is not used here at all

File: deeppavlov/dataset_readers/ubuntu_dstc7_mt_reader.py [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+                for i, utterance in enumerate(dialog['options-for-next']):
+                    if utterance['candidate-id'] != target_id:
+                        fake_responses.append(utterance['utterance'])
+
+                # aligned list of context utterances
+                expanded_context = self._expand_context(utterances, padding=self.padding)
+
+                if mode == 'train':
+                    data.append((expanded_context + [true_response], 1))
+                    data.append((expanded_context + list(np.random.choice(fake_responses, size=1)), 0))  # random 1 from 99
```
Comment (yoptar) [inline deeppavlov/dataset_readers/ubuntu_dstc7_mt_reader.py]: Shouldn't there be an ability to set a random seed? Something like
```python3
self.np_random = np.random.RandomState(seed)
...
self.np_random.choice(fake_responses, size=1)
```

File: deeppavlov/models/ranking/deep_attention_matching_network_use_transformer.py [hunk 11]
```diff
@@ -0,0 +1,410 @@
...
+        decay_steps (int): Number of steps after which is to decay the learning rate.
+    """
+
+    def __init__(self,
+                 batch_size: int,
+                 embedding_dim: int = 200,
+                 num_context_turns: int = 10,
+                 max_sequence_length: int = 50,
+                 learning_rate: float = 1e-3,
+                 emb_matrix: np.ndarray = None,
```
Comment (yoptar) [inline deeppavlov/models/ranking/deep_attention_matching_network_use_transformer.py]:
```suggestion
                 emb_matrix: Optional[np.ndarray] = None,
```

File: deeppavlov/models/ranking/deep_attention_matching_network_use_transformer.py [hunk 12]
```diff
@@ -0,0 +1,410 @@
...
+                 trainable_embeddings: bool = False,
+                 is_positional: bool = True,
+                 stack_num: int = 5,
+                 seed: int = 65,
+                 decay_steps: int = 600,
+                 *args,
+                 **kwargs):
+
+        self.seed = seed
+        np.random.seed(self.seed)
```
Comment (yoptar) [inline deeppavlov/models/ranking/deep_attention_matching_network_use_transformer.py]: Again, don't see `np.random` used anywhere in this class

File: deeppavlov/models/ranking/matching_models/dam_utils/operations.py [hunk 13]
```diff
@@ -0,0 +1,395 @@
...
+# }
+# ```
+# http://aclweb.org/anthology/P18-1103
+#
+# Based on authors' Tensorflow code: https://github.com/baidu/Dialogue/tree/master/DAM
+
+from logging import getLogger
+
+import math
+import numpy as np
```
Comment (yoptar) [inline deeppavlov/models/ranking/matching_models/dam_utils/operations.py]: `math` module is from the standard library so it should be moved higher up

File: deeppavlov/models/ranking/tf_base_matching_model.py [hunk 14]
```diff
@@ -0,0 +1,163 @@
...
+        batch_buffer_response_len = []  # [batch_size]
+
+        context_sentences = sample[:self.num_context_turns]
+        response_sentences = sample[self.num_context_turns:]
+
+        # Format model inputs:
+        # 4 model inputs
+
+        # 1. Token indices for context
+        batch_buffer_context += [context_sentences for sent in response_sentences]
```
Comment (yoptar) [inline deeppavlov/models/ranking/tf_base_matching_model.py]:
```suggestion
        batch_buffer_context += [context_sentences] * len(response_sentences)
```

File: deeppavlov/models/ranking/tf_base_matching_model.py [hunk 15]
```diff
@@ -0,0 +1,163 @@
...
+        # 2. Token indices for response
+        batch_buffer_response += [response_sentence for response_sentence in response_sentences]
```
Comment (yoptar) [inline deeppavlov/models/ranking/tf_base_matching_model.py]:
```suggestion
        batch_buffer_response += list(response_sentences)
```

File: deeppavlov/models/ranking/tf_base_matching_model.py [hunk 16]
```diff
@@ -0,0 +1,163 @@
...
+        # 3. Lens of context sentences
+        lens = []
+        for context in [context_sentences for sent in response_sentences]:
```
Comment (yoptar) [inline deeppavlov/models/ranking/tf_base_matching_model.py]:
```suggestion
        for context in batch_buffer_context:
```

File: deeppavlov/models/ranking/tf_base_matching_model.py [hunk 17]
```diff
@@ -0,0 +1,163 @@
...
+            context_sentences_lens = []
+            for sent in context:
+                context_sentences_lens.append(len(sent[sent != 0]))
+            lens.append(context_sentences_lens)
+        batch_buffer_context_len += lens
+        # 4. Lens of context sentences
+        lens = []
+        for context in [response_sentence for response_sentence in response_sentences]:
```
Comment (yoptar) [inline deeppavlov/models/ranking/tf_base_matching_model.py]:
```suggestion
        for context in response_sentences:
```

File: deeppavlov/dataset_readers/ubuntu_v1_mt_reader.py [hunk 18]
```diff
@@ -0,0 +1,119 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import List, Tuple, Union, Dict
+from pathlib import Path
+import codecs
```
Comment (yoptar) [inline deeppavlov/dataset_readers/ubuntu_v1_mt_reader.py]:
```suggestion
```
